orderer/common/follower/follower_chain.go

```diff
@@ -426,6 +426,129 @@ func (c *Chain) pullUpToJoin() error {
+		configBlock, err := c.pullUntilNextConfig()
+		if err != nil {
+			return errors.Wrap(err, "failed to pull until next config")
+		}
+
+		channelMember, err = c.clusterConsenter.IsChannelMember(configBlock)
+		if err != nil {
+			return errors.Wrap(err, "failed to determine channel membership from pulled config")
+		}
+
+		c.logger.Debugf("next config block: %d, channel member: %v", c.lastConfig.Header.Number, channelMember)
+	}
+
+	return err
+}
+
+// pullUntilNextConfig will return the next config block or an error.
+func (c *Chain) pullUntilNextConfig() (*common.Block, error) {
+	// This puller is built from the tip of the ledger. When we get a new config block we exit the method. That will
+	// destroy the puller. Next time we enter, a new puller is built, taking in the new config.
+	puller, err := c.createBlockPullerFunc()
+	if err != nil {
+		c.logger.Errorf("Error creating block puller: %s", err)
+		return nil, errors.Wrap(err, "error creating block puller")
+	}
+	defer puller.Close()
+
+	var configBlock *common.Block
+	for configBlock == nil {
+		endpoint, latestHeight, err := cluster.LatestHeightAndEndpoint(puller)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get latest height and endpoint")
+		}
+		c.logger.Debugf("Orderer endpoint %s has the biggest ledger height: %d", endpoint, latestHeight)
+
+		if latestHeight <= c.ledgerResources.Height() {
+			c.logger.Debugf("My height: %d, latest height: %d; going to wait %v for latest height to grow",
+				c.ledgerResources.Height(), latestHeight, c.retryInterval)
+			select {
+			case <-c.stopChan:
+				return nil, errors.New("stopped while waiting for latest height to grow")
+			case <-c.timeAfter(c.retryInterval):
+				c.increaseRetryInterval()
+				continue
+			}
+		}
+
+		configBlock, err = c.pullUntilLatestOrConfig(puller, latestHeight)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to pull until latest height or config")
+		}
+		c.resetRetryInterval()
+	}
+
+	return configBlock, nil
+}
+
+func (c *Chain) pullUntilLatestOrConfig(puller ChannelPuller, latestHeight uint64) (*common.Block, error) {
+	firstBlockToPull := c.ledgerResources.Height()
+	var nextBlock = c.ledgerResources.Block(firstBlockToPull - 1)
+	if nextBlock == nil {
+		return nil, errors.Errorf("cannot retrieve previous block %d", firstBlockToPull-1)
+	}
+	var actualPrevHash = protoutil.BlockHeaderHash(nextBlock.Header)
+
+	// Pull until the latest height or a config block
+	for seq := firstBlockToPull; seq < latestHeight; seq++ {
+		select {
+		case <-c.stopChan:
+			c.logger.Warnf("Stopped while pulling blocks: from %d until %d, last pulled %d",
+				firstBlockToPull, latestHeight, nextBlock.Header.Number)
+			return nil, errors.New("stopped while pulling blocks")
+		default:
+			nextBlock = puller.PullBlock(seq)
+			if nextBlock == nil {
+				return nil, errors.Wrapf(cluster.ErrRetryCountExhausted, "failed to pull block %d", seq)
+			}
+			reportedPrevHash := nextBlock.Header.PreviousHash
```

Comment:
I'm not strictly opposed to these checks, but I'm trying to understand the adversarial model.

If the adversary is malicious, then it's trivial to report a correct previous block hash, and get past this check.

If the adversary is not malicious, and this is a simple namespace intersection, then I would generally expect the signature checking to fail.

The ledger code will not allow blocks which do not chain together to be persisted (it results in a panic).

Reply:
I had this discussion with @yacovm  in the previous PR, see:
https://github.com/hyperledger/fabric/pull/1536#discussion_r450500840

Let me sum it up. The BlockPuller works by fetching a batch of blocks from another orderer and verifying that batch. It verifies the hash chain within a batch, but there is no check for the continuity of hashes between batches. Moreover, not all the signatures in a batch are verified. 

I think this needs to be fixed at the cluster package under a different story (see link above for more details). 

Exactly because an orderer panics, we want to be careful with data we get from external parties before we submit it to the ledger. We better have a follower (channel) not making progress than crashing the entire orderer.

Reply:
I think the problem Yoav is trying to protect against, is that an orderer has a block with intact block data and header, but incorrect signatures due to storage corruption. 
This orderer isn't aware the signatures of its block are corrupt and can silently replicate this block to other orderers via either Raft replication or onboarding.
Further assume that all orderers with intact signatures on that block are removed from the channel, and you get a channel with only orderers that have corrupt signatures on that block, and then peers that join the network and only replicate from orderers, are stuck.

One way of solving this, is to check signature in every block. 
Another way would be to have the storage layer (the common ledger) use methods to detect corruption of the block when it reads it from the file system.

Reply:
@yacovm I don't think this check addresses anything with signatures of metadata? It is only checking the block header hash, which, is unconnected to the metadata?

Assuming that we have some correct signature over a future block which chains to this one, how do we get a hash mismatch to what's on our chain? The only way I see, is if someone accidentally re-used crypto material and had a channel ID conflict.  Basically, two blockchains with the same orderers, with the same name, but different contents.  We can protect against this... just trying to understand the point of the check.

Reply:
> @yacovm I don't think this check addresses anything with signatures of metadata? It is only checking the block header hash, which, is unconnected to the metadata?

I was never referring to the code, but to what Yoav wrote in his comment:
>  Moreover, not all the signatures in a batch are verified.

> Assuming that we have some correct signature over a future block which chains to this one, how do we get a hash mismatch to what's on our chain?

We don't get a hash mismatch... I am saying that if you don't check every signature then if there is an orderer that has a non-crash type of error (corrupt signature but intact block), under certain conditions and execution, this corrupt signature will propagate to all nodes in the system.

Reply:
Okay.  So, we are discussing two things here:

1. In the code that this thread references, we are checking that the previous block hash matches the block that is committed onto our ledger.
2. In the block replication code, for efficiency, we are not checking every signature, but rather only checking the last signature, and then verifying the hash chain.  But, this will not detect corruption in metadata.

For (1), the situation this code is defending against _should not_ ever occur.  As it implies two different valid blocks of the same sequence number (as they must both be covered by a valid signature, either directly or by chaining).  And we have a fork in our blockchain.  It might still be worth defending against, because we would rather not have the whole orderer crash, but importantly, this is not a security check, as security guarantees have already been violated if this check fails.  Instead, this is more of a sanity check in case of operator incompetence.

For (2), this does seem like a problem of sorts.  In the CFT case, it requires corruption, but in the eventual BFT case, obviously we'd need something more robust.  But as @tock-ibm says, an issue to be handled separately from this series.

Reply:
For (2) I suggest that we solve this problem universally in the common ledger by appending [CRC](https://golang.org/pkg/hash/crc32/) to every block we write into the filesystem and when we retrieve it, we validate the tag matches the CRC of the block.

Reply:
opened https://jira.hyperledger.org/browse/FAB-18104 for this issue.